misago/legal/models.py

```diff
@@ -1 +1,127 @@
+    PRIVACY = 'privacy_policy'
+    Legal = (
+        (TOS, 'Terms of service'),
+        (PRIVACY, 'Privacy Policy'),
+    )
+    active = models.BooleanField(
+        default=False,
+        verbose_name=_('active'),
+        help_text=_(
+            u'Only one terms of service is allowed to be active'
+        )
+    )
+    title = models.CharField(
+        max_length=20,
+        choices=Legal,
+        default=TOS
+    )
+    published = models.DateTimeField(editable=False, null=True)
+    importance = models.IntegerField(
+        choices=IMPORTANCE,
+        default=IMP_HIGH,
+    )
+    text = models.TextField(verbose_name=_('content'), blank=True)
+    objects = LegalAgreementManager()
+
+    def latest_published(self):
+        return LegalAgreementVersion.objects.filter(title=self.title).order_by('-version')
+
+    def save(self, *args, **kwargs):
+        """ Ensure that only LegalAgreement is active"""
+        if self.active:
+            LegalAgreement.objects.exclude(id=self.id).update(active=False)
+        else:
+            if not LegalAgreement.objects \
+                    .exclude(id=self.id) \
+                    .filter(active=True, title=self.title) \
+                    .exists():
+                raise NoActiveLegalAgreement(
+                    u'One of the {} must be marked active'.format(self.title)
+                )
+
+        if self.created == None:
+            self.created = datetime.datetime.utcnow()
+        super(LegalAgreement, self).save(*args, **kwargs)
+        if self.active:
+            latest_published = self.latest_published()
+            if not latest_published or self.text != latest_published[0].text:
+                self.published = datetime.datetime.utcnow()
+                newVersion = LegalAgreementVersion(legal_agreement=self, title=self.title, text=self.text,
+                                                   created=self.created, published=self.published)
+                newVersion.save()
+
+
+class LegalAgreementVersion(BaseModel):
+    legal_agreement = models.ForeignKey(LegalAgreement)
+    version = models.IntegerField(null=True)
+    title = models.CharField(max_length=20)
+    text = models.TextField()
+
+    class Meta:
+        unique_together = ('legal_agreement', 'version')
+
+    def save(self, *args, **kwargs):
+        # Count each release.
+        current_version = LegalAgreementVersion.objects.filter(title=self.title).order_by('-version')[:1]
+        self.version = current_version[0].version + 1 if current_version else 1
+        super(LegalAgreementVersion, self).save(*args, **kwargs)
+
+
+class UserAgreement(BaseModel):
+    legal_agreement = models.ForeignKey(LegalAgreement, related_name='legal_agreement')
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        blank=True,
+        null=True,
+        related_name='user_agreement'
+    )
+    user_ip = models.GenericIPAddressField()
+    accept = models.BooleanField(default=False)
```

Comment:
How do we know what version of legal agreement user has accepted, and when? How do we plan to handle the situation when user opts in and out of agreement repeatedly?